### PR TITLE
Provide module Ki.Typed to prevent escape

### PR DESCRIPTION
**Pull request by LSLeary**
### A summary in one type signature

```haskell
scoped :: (forall s. Scope s -> IO a) -> IO a
```

### Commit: Provide module Ki.Typed to prevent escape

> RankNTypes encode the notion of an inviolable *scope* at the type level.
> As `runST` famously demonstrates, this can be utilised to disallow
> value-level scope violations at compile time.
>
> `Scope` and `Thread` are invalid outside their parent scope, hence a new
> interface is provided to enforce a scoping invariant. However, in order
> for the internals to leverage that invariant, the original interface
> must go.

### Disclaimer

  - The entire PR is predicated upon the notion that the internals could actually leverage the invariant to significant effect, but I'm relying on the expertise of the mainters to determine whether it's so. If not, the PR should be rejected.

  - I only discovered "structured concurrency" this morning; I'm operating from a position of relative ignorance.

### CODE REVIEW

**Comment by mitchellwrosen** [comment]
Hi, thanks for the PR. At this time I'm not inclined to merge, but please feel free to make a [discussion](https://github.com/awkward-squad/ki/discussions) about the API where others can chime in.

Rationale:

- That certain "resources" are only valid within a certain lexical scope is very common. `Scope` is one of those resources, but so are `Connection` things, `Handle` things, and so on. So, Haskellers (and, well, all programmers) are already used to taking some care not to violate the lifecycle of such resources by using them outside the scope in which they are valid. Therefore, I don't believe one is likely to even ever attempt to run e.g. `scoped pure` accidentally. 

- The `ST` monad indeed uses this trick to prevent `STRef` leakage. But if an `STRef` leaks, we can violate referential transparency. There's no such danger here - if one tries to return a `Scope` after it's been closed, they will simply get a runtime exception (`error` call).

In short: complicating the interface in this way doesn't seem worth the bump in complexity for users, in terms of both understanding and using the API.

Anecdotally, I've not once seen a resource with a phantom type variable like the one you propose on `Scope` - just out of curiosity, do you happen to have an example from some library on Hackage?

**Comment by LSLeary** [comment]
Sorry for the delay; I got stuck on another project.

Re scoped "resources", haskellers are a bit different—we're constantly constructing and passing around both functions and actions as first class values, and putting them in data structures too.

I think it's actually quite easy to accidentally let something escape when it's internal to some action that's internal to some data. And, in fact, I did exactly that just the other day, in the aforementioned project.

I don't want to be digging around for the precise cause of an `error` call after a runtime failure when a type error could have given me a location before the thing even ran, nor do I want ki to slow down my program by checking for these scope violations itself.

If we can make ki safter and faster while preserving the value-level interface, and only complicating the type-level interface a little, then I think it's well worth it.

---

Well, that's what I wanted to say, anyway. Unfortunately, in the intervening days, I noticed a couple of problems with my approach. I'd thought I could avoid introducing a monad, but I was wrong.

The first issue is that functions like

```haskell
fork_ :: Scope s -> IO Void -> IO ()
```

are eliminators for `s`. You can't write `scoped pure`, but you *can* write `scoped (pure . fork_) :: IO (IO Void -> IO ())`, which will sneak out a `Scope`.

To prevent this, we must also stick `s` to the resulting `IO` in an opaque newtype.

```haskell
newtype Scoped s f a = Scoped{ runScoped :: ReaderT Ki.Scope f a }
```

The interface then becomes:

```haskell
scoped :: (forall s. Scoped s IO a) -> IO a

fork :: Scoped s IO a -> Scoped s IO (Thread s a)

await :: Thread s a -> Scoped s STM a
```

Etc. This is a substantial change to the interface, both at the value and the type level, but it also has the benefit that `Scope` is no longer visible—users don't even need to know it exists, let alone explicitly thread it through their code.

Unfortunately, this is where the next problem arises: `Scoped s` cannot safely support unlifting; it cannot have `MonadUnliftIO`. If we were to expose

```haskell
withRunScopedInF :: ((forall a. Scoped s f a -> f a) -> f b) -> Scoped s f b
```

then it would again be possible to write, e.g.,

```haskell
sneakOutFork_ :: IO (IO Void -> IO ())
sneakOutFork_ = scoped $ withRunScopedInF \run ->
  pure (run . fork_ . lift)
```

For a monad that ought to be a thin wrapper over `IO`, failure to support unlifting is a dealbreaker for many users, so I guess this is dead in the water. Re examples, perhaps this is why I haven't seen the same approach on on Hackage; it's difficult to provide powerful primitives like `withRunScopedInF` without creating backdoors.

*This comment and the associated force-push are for posterity.*
